oresat_configs/edl_cmd_defs.py

```diff
@@ -0,0 +1,227 @@
+                fmt = _COMMAND_DATA_FMT[f.data_type]
+                data[f.name] = struct.unpack(fmt, tmp)[0]
+            elif f.data_type == "bytes":
+                if f.size_prefix != "":  # dynamic length in bits
+                    size_prefix = int(f.size_prefix[4:]) // 8
+                    data_type_size_raw = raw[offset : offset + size_prefix]
+                    data_type_size = int.from_bytes(data_type_size_raw, "little") // 8
+                    offset += size_prefix
+                else:  # fix_size
+                    data_type_size = f.fixed_size
+                data[f.name] = raw[offset : offset + data_type_size]
+            elif f.data_type == "str":
+                if f.max_size != "":  # dynamic length that ends with "\0"
+                    data_type_size = raw[offset:].find(b"\0")
+                else:  # fix_size
+                    data_type_size = f.fixed_size
+                data[f.name] = raw[offset : offset + data_type_size].decode()
+            else:
+                raise ValueError(f"invalid edl field {f.name}")
+            offset += data_type_size
+
+        return tuple(data.values())
+
+    def _encode(self, values: tuple[Any], fields: list[EdlCommandField]) -> bytes:
+
+        if not isinstance(values, (tuple, list)):
+            values = (values,)
+
+        if len(values) != len(fields):
+            raise ValueError(
+                f"invalid number of values for packet: got {len(values)} expected {len(fields)}"
+            )
+
+        # fixed size packet - quick encode
+        if not self._dynamic_len(fields):
+            fmt = "".join([_COMMAND_DATA_FMT[f.data_type] for f in fields])
+            return struct.pack(fmt, *values)
+
+        # dynamic size packet - slower encode
+        raw = b""
+        for f, v in zip(fields, values):
+            if f.data_type in _COMMAND_DATA_FMT:
+                fmt = _COMMAND_DATA_FMT[f.data_type]
+                raw += struct.pack(fmt, v)
+            elif f.data_type == "bytes":
+                value = v
+                if f.size_prefix != "":  # dynamic length in bits
+                    fmt = _COMMAND_DATA_FMT[f.size_prefix]
+                    raw += struct.pack(fmt, len(v) * 8)
+                else:  # fixed length
+                    value += b"\x00" * (f.fixed_size - len(value))
+                raw += value
+            elif f.data_type == "str":
+                value = v.encode()
+                if f.max_size != "":  # dynamic length that ends with "\0"
+                    value += b"\0"
+                else:  # fixed length
+                    value += b"\0" * (f.fixed_size - len(value))
+                raw += value
+            else:
+                raise ValueError(f"invalid data type {f.data_type} for edl field {f.name}")
+        return raw
+
+    def decode_request(self, raw: bytes) -> tuple[Any]:
+        """Decode a EDL request payload."""
+        return self._decode(raw, self.request)
```

Comment:
I'm having a hard time grokking why self.request is embedded into the object but values are not.


Is this object intended to keep-state on anything?

Reply:
Nope, it is just the definition